### PR TITLE
Allow updating Client Cert w/o the private key

### PR DESCRIPTION
**Pull request by talanknight**

### CODE REVIEW

File: internal/credential/vault/repository_credential_store.go [hunk 1]
```diff
@@ -499,6 +487,50 @@ func (ps *privateStore) TableName() string {
 	return "credential_vault_store_private"
 }
 
+// applyUpdate takes the new and applies it to the orig using the field masks
+func applyUpdate(new, orig *CredentialStore, fieldMaskPaths []string) *CredentialStore {
```
Comment (mgaffney) [inline internal/credential/vault/repository_credential_store.go:491]: Why not make this a method on `*CredentialStore`?

File: internal/credential/vault/repository_credential_store.go [hunk 2]
```diff
@@ -718,50 +710,78 @@ func (r *Repository) UpdateCredentialStore(ctx context.Context, cs *CredentialSt
...
+				}
+				token, err = newToken(cs.GetPublicId(), cs.inputToken, []byte(accessor), tokenExpires)
+				if err != nil {
+					return err
+				}
+				// encrypt token
+				if err := token.encrypt(ctx, databaseWrapper); err != nil {
+					return errors.Wrap(err, op)
+				}
+				query, values := token.insertQuery()
```
Comment (mgaffney) [inline internal/credential/vault/repository_credential_store.go:776]: Why move all of this within the scope of the transaction? In general, we should try to keep the time of a database transaction as short as possible?

**Review by mgaffney — CHANGES_REQUESTED** [review]
Overall, this looks good but I'm concerned about the amount of logic moved to be within the scope of a database transaction (especially the parts with external calls being made). Can you take another look and try to move that out of the database transaction?

**Review by louisruch — APPROVED** [review]
This LGTM!

File: internal/credential/vault/credential_store.go [hunk 3]
```diff
@@ -97,3 +97,20 @@ func (cs *CredentialStore) Token() *Token {
...
+		TlsServerName: cs.TlsServerName,
+		TlsSkipVerify: cs.TlsSkipVerify,
+		Namespace:     cs.Namespace,
+	}
+	if cs.clientCert != nil {
+		clientConfig.ClientCert = cs.clientCert.GetCertificate()
+		clientConfig.ClientKey = cs.clientCert.GetCertificateKey()
+	}
+
+	return newClient(clientConfig)
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/credential/vault/credential_store.go]: Yes, thank you.

File: internal/credential/vault/repository_credential_store.go (hunk 1 above)
Comment (talanknight) [reply internal/credential/vault/repository_credential_store.go]: Done, thanks!

File: internal/credential/vault/repository_credential_store.go (hunk 2 above)
Comment (talanknight) [reply internal/credential/vault/repository_credential_store.go]: Moved a bunch of stuff outside of the transaction.

File: internal/credential/vault/repository_credential_store.go [hunk 4]
```diff
@@ -499,6 +487,50 @@ func (ps *privateStore) TableName() string {
 	return "credential_vault_store_private"
 }
 
+// applyUpdate takes the new and applies it to the orig using the field masks
+func applyUpdate(new, orig *CredentialStore, fieldMaskPaths []string) *CredentialStore {
+	cp := orig.clone()
```
[the remark this answers is not in the document]
Comment (talanknight) [reply internal/credential/vault/repository_credential_store.go]: cp was abbreviated for copy, happy to rename if you want.

[the remark this answers is not in the document]
Comment (louisruch) [reply internal/credential/vault/repository_credential_store.go]: Aha! No this is good thanks

File: internal/credential/vault/repository_credential_store.go [hunk 5]
```diff
@@ -533,68 +521,73 @@ func (r *Repository) UpdateCredentialStore(ctx context.Context, cs *CredentialSt
...
-		case strings.EqualFold("Token", f):
+		case strings.EqualFold(nameField, f):
+		case strings.EqualFold(descriptionField, f):
+		case strings.EqualFold(namespaceField, f):
+		case strings.EqualFold(tlsServerNameField, f):
+		case strings.EqualFold(tlsSkipVerifyField, f):
+		case strings.EqualFold(caCertField, f):
+		case strings.EqualFold(vaultAddressField, f):
+		case strings.EqualFold(certificateField, f):
+		case strings.EqualFold(certificateKeyField, f):
```
Comment (mgaffney) [inline internal/credential/vault/repository_credential_store.go:535]: Can you update the method doc regarding the ability to update the client certificate without updating the client certificate private key?

**Review by mgaffney — APPROVED** [review]
Looks great! Just some minor comments about missing docs and potentially a method doc that needs to be updated.

File: internal/credential/vault/repository_credential_store.go (hunk 2 above)
Comment (mgaffney) [reply internal/credential/vault/repository_credential_store.go]: ty

**Pull request merged**